canonical/maas-ui
# refactor(machines): cache machine list selectors

## Done

- refactor(machines): cache machine list selectors
- add `createCachedSelector` from `re-reselect` for proper memoization when using different systemId as a paramter

This results in significantly improved performance for large number of machines, in particular whenever there's an update that does not affect any of the machines.

update render is on average **~40x faster** when rendering 200 machines (down to 1ms from 40ms)

## QA

1. Open to machine list
2. Open developer tools
3. Enable highlighting updates on component render in React devtools by going to Components tab, clicking the cog icon and clicking "Highlight updates when components render."
4. Start React profiler by going to the profiler tab and clicking the Record icon
5. Open Redux Developer tools
6. Dispatch an action that will affect state but is unrelated to the list of machines, e.g.
```
{
  type: 'service/createNotify',
  payload: {
    id: 999,
    name: 'dhcpd',
    status: 'dead',
    status_info: ''
  }
}
```
7. Wait for the page to have some changes highlighted by the React devtools
8. Stop the profiler
9. Verify that the subsequest machine list update has a value that's lower than on the main branch (~ 1ms instead of ~40ms)

### MAAS deployment

To run this branch you will need access to one of the following MAAS deployments:

- [Karura](/HACKING.md#karura)
- [Bolla](/HACKING.md#bolla)
- [A development MAAS](/HACKING.md#development-deployment)
- [A local snap MAAS](/HACKING.md#snap-deployment) (this will not usually have machines)

### Running the branch

You can run this branch by:

- Serving with [dotrun](/HACKING.md#maas-ui-development-setup)
- [Building in a development MAAS](/HACKING.md#running-maas-ui-from-a-development-maas)


## Fixes

Fixes: [MAASENG-1419](https://warthogs.atlassian.net/browse/MAASENG-1419)

## Launchpad issue

Related Launchpad maas issue in the form `lp#number`.

## Backports

In general, please propose fixes against _main_ rather than release branches (e.g. 2.7), unless the fix is only applicable for that specific release. Please apply backport labels to the PR (e.g. "Backport 2.7") for the appropriate releases to target.

Only bug and security fixes should be backported, new features should only land in main.

## Screenshots

### before
**MachineList component tree render ~40ms**

![image](https://user-images.githubusercontent.com/7452681/224014270-c0539982-54d2-4f3d-b0fa-ca508295f4fd.png)

### after
**MachineList component tree render ~1ms**

![image](https://user-images.githubusercontent.com/7452681/224014296-9442617e-9f9b-449e-95f4-ab8ff3405933.png)


[MAASENG-1419]: https://warthogs.atlassian.net/browse/MAASENG-1419?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ
